### PR TITLE
Check for updated bgg game info

### PR DESCRIPTION
**Pull request by JamesRichards07**
1. The check for bgg updates feature works as best as I can tell. I wasn't sure how to test this since the information is coming for the bgg api.
2. The execute method is what checks for any new information. We had discussed making the ``save`` method do that but I wasn't sure of any easy way to do this.
3. I update the ``$me`` variable in the ``save`` method to just ``source``.

### CODE REVIEW

File: src/Sources/AbstractSource.php [hunk 1]
```diff
@@ -135,24 +135,25 @@ function ($migration_class) {
...
     {
-        // @todo: update $me name
         // @todo: Only save model if something has changed
         // @todo: in the Audit, describe what changed
+
         // Now, we save the primary version
-        $primary->save();
+        if($primary !== null){
+            $primary->save();
+            $source->game_id = $primary->id;
```
Comment (electricjones) [inline src/Sources/AbstractSource.php:146]: Yeah, this is something I did that was wrong. We'll have to figure something out here.

`game_id` works for us, but that won't be the column name for others. And the framework can't care about those things.

So, I think we have 2 options:
1. Require the source itself to populate this column
2. Require the source to declare what the column name is. Add another `abstract function getPrimaryEntityJoinColumnName()` or something like that. Then, here we can 

```php
$column_name = static::getPrimaryEntityJoinColumnName();
$source->{$column_name} = $primary->id;
```

File: src/Ravenfire/BoardGameGeek/BoardGameGeekSource.php [hunk 2]
```diff
@@ -153,21 +119,87 @@ public function saveGameInfo(string $gameId, $client)
...
+            $game = null;
+        };
+
+        $boardgamegeek->bgg_foreign_id = $gameId;
+        $this->createNewBoardGameGeek($response, $bggColumns, $boardgamegeek);
+
+        if(count($boardgamegeekgame) !== 0){
+            foreach($bggColumns as $bggColumn){
+                if($boardgamegeekgame[0]->$bggColumn != $boardgamegeek->$bggColumn){
+                    dd($boardgamegeekgame[0]->$bggColumn, $boardgamegeek->$bggColumn);
```
Comment (electricjones) [inline src/Ravenfire/BoardGameGeek/BoardGameGeekSource.php:168]: I know this is a draft. Just a friendly reminder not to forget about this `dd()`

File: src/Sources/AbstractSource.php [hunk 3]
```diff
@@ -135,24 +135,25 @@ function ($migration_class) {
...
         // Create new record in this source
-        $me->save();
+        $source->save();
 
         // Create a new Audit Record because something changed
         $audit = new Audit();
         $audit->job_id = $this->getJob()->id;
-        $audit->record_id = $me->id;
+        $audit->record_id = $source->id;
         $audit->source_key = static::getKey();
```
Comment (electricjones) [inline src/Sources/AbstractSource.php:157]: Here is where we will (eventually) want to add more.

```
foreach ($what_changed as $change) {
       $audit = new Audit();
        $audit->job_id = $this->getJob()->id;
        $audit->record_id = $source->id;
        $audit->source_key = static::getKey();
        $audit->column = $change['column'];
        $audit->previous = $change['previous'];
        $audit->new = $change['new'];
}
```

OR

```php
       $audit = new Audit();
        $audit->job_id = $this->getJob()->id;
        $audit->record_id = $source->id;
        $audit->source_key = static::getKey();
        $audit->changes = json_encode($what_changed); 
```

In both cases `$what_changed` will be an array with the column that changed, the old value, and the new value

**Comment by electricjones** [comment]
Yeah, I definitely see your dilemma. I didn't think that far through.

I think I see two open questions/ issues here:
1. When to save the `game` because we want that to be in `save()` but it really can't be.
2. How/when to determine when something has changed.

I think you have both solutions here -- we can just clean them up a bit. Mostly by pulling what you have into a few more methods.

Let's talk about the second one for now. You have a great way to tell if something has changed. You build up an array of columns to look at and then check each api response against the model you have.

This can be extracted down into `AbstractSource` and moved into save. Though, we'll have to modify `save()` or `AbstractSource` a bit. You can either pass in the columns array as an argument or add an abstract method to `AbstractSource` and require the source to declare that array.

In either case, the `AbstractSource` now has everything it needs. It can cycle through the array w/o caring what's in it, compare the data, and know if something needs saved.

Furthermore, you can actually know **what** needs saved.

So, the common thing here is to do something like this:

```php
function findChanges($columnsToLookAt, $existingModel, $newData): array {
    // Do basically what you are doing but instead of returning a `bool`, return an array.
    // An empty array just means there are no changes
}
```
That also allows you to do what I suggested in the `save()` method with the audit table and save **what** actually changed.

**Comment by electricjones** [comment]
But, we could take it even a step further.

My original thought is that the `execute()` methods would just construct two objects a `game` and a `board_game_geek`. Each object would be a model. The execute() method doesn't need to grab anything from the database to do this. It just:

```
$game = new Game();
$game->whatever = $response['stuff'];
```

And so on.

Then, its the save method that checks to see if it exists, and if it does exist, finds what's changed and saves everything.

That may have been naive. That may not be as possible as I thought, but I think we can get closer.

File: src/Sources/AbstractSource.php (hunk 1 above)
Comment (JamesRichards07) [reply src/Sources/AbstractSource.php:146]: So why can't the other sources also have a ``game_id`` column? Isn't that what a one-to-many looks like?

Comment (electricjones) [reply src/Sources/AbstractSource.php:146]: It's not that the other **sources** can't. In fact, they will.

It's that other **projects** won't. If the framework is used to collect financial data for car sales, the primary entity will be something like `cars` and each source will want to have a `car_id`, not a `game_id`

Everything in `AbstractSource` needs to be blind to what kind of project this is.

**Pull request merged**